node-gui/src/backend/wallet_events/mod.rs

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2023 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use tokio::sync::Notify;
+use wallet::wallet_events::WalletEvents;
+use wallet_types::{AccountWalletTxId, WalletTx};
+
+pub struct GuiWalletEvents {
+    wallet_notify: Arc<Notify>,
+    updated: bool,
+}
+
+impl GuiWalletEvents {
+    pub fn new(wallet_notify: Arc<Notify>) -> Self {
+        GuiWalletEvents {
+            wallet_notify,
+            updated: false,
+        }
+    }
+
+    fn notify(&mut self) {
+        self.updated = true;
+        self.wallet_notify.notify_one();
+    }
+
+    pub fn is_set(&self) -> bool {
+        self.updated
+    }
+
+    pub fn reset(&mut self) {
+        self.updated = false;
+    }
+}
+
+impl WalletEvents for GuiWalletEvents {
+    fn new_block(&mut self) {
```

Comment:
From the looks of it, these functions shouldn't be mutable. It only complicates things. It seems to me that the right approach is to use some kind of channels to isolate this mutability.

Reply:
I tried to use channels but I did not like the result. The main problem with channels is that we only need to update the UI once, but with channels we can get many duplicate events.
I don't see the problem with mutability.

Reply:
I switched to channels and made the `WalletEvents` trait non-mutable.